quantities/elementary_functions_body.hpp

```diff
@@ -5,30 +5,38 @@
 #include <pmmintrin.h>
 
 #include <cmath>
-#include <type_traits>
 
+#include "boost/multiprecision/cpp_int.hpp"
 #include "numerics/cbrt.hpp"
 #include "numerics/fma.hpp"
 #include "numerics/next.hpp"
+#include "quantities/concepts.hpp"
 #include "quantities/si.hpp"
 
 namespace principia {
 namespace quantities {
 namespace _elementary_functions {
 namespace internal {
 
+using namespace boost::multiprecision;
 using namespace principia::numerics::_cbrt;
 using namespace principia::numerics::_fma;
 using namespace principia::numerics::_next;
+using namespace principia::quantities::_concepts;
 using namespace principia::quantities::_si;
 
 template<typename Q1, typename Q2>
 Product<Q1, Q2> FusedMultiplyAdd(Q1 const& x,
                                  Q2 const& y,
                                  Product<Q1, Q2> const& z) {
-  return si::Unit<Product<Q1, Q2>> *
-         numerics::_fma::FusedMultiplyAdd(
-             x / si::Unit<Q1>, y / si::Unit<Q2>, z / si::Unit<Product<Q1, Q2>>);
+  if constexpr (is_number<Q1>::value || is_number<Q2>::value) {
+    return x * y + z;
```

Comment:
This condition should also check for `number_category<Q2>::value == number_kind_integer || number_category<Q2>::value == number_kind_rational`; for floating-point we cannot just do that (but I think there is an `fma`).